java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java

```diff
@@ -106,28 +105,134 @@ private static boolean isExcluded(MethodTree tree) {
+    }
     Symbol symbol = tree.symbol();
-    if (!atLeastOneReference.isEmpty() && !atLeastOneReference.peek() && !currentMethod.peek().equals(symbol)  && referenceInstance(symbol)) {
-      atLeastOneReference.pop();
-      atLeastOneReference.push(Boolean.TRUE);
+    final MethodReference currentMethod = methodReferences.peek();
+    if (symbol.isUnknown()) {
+      currentMethod.setNonStaticReference();
+      return;
+    }
+    for (MethodReference methodReference : methodReferences) {
+      methodReference.checkSymbol(symbol);
     }
   }
 
-  private boolean referenceInstance(Symbol symbol) {
-    return symbol.isUnknown() || (!symbol.isStatic() && fromInstance(symbol.owner()));
+  @Override
+  public void visitMemberSelectExpression(MemberSelectExpressionTree tree) {
+    if (tree.expression().is(Tree.Kind.IDENTIFIER)) {
+      IdentifierTree identifier = (IdentifierTree) tree.expression();
+      final Symbol owner = identifier.symbol().owner();
+      if (owner != null && owner.isMethodSymbol()) {
+        // No need to investigate selection on local symbols
+        return;
+      }
+    }
+    if (tree.expression().symbolType().isSubtypeOf("java.lang.Class")) {
+      // No need to investigate selection on a Class object
+      return;
+    }
+    super.visitMemberSelectExpression(tree);
   }
 
-  private boolean fromInstance(Symbol owner) {
-    for (Symbol outerClass : outerClasses) {
-      Type ownerType = owner.type();
-      if (ownerType != null) {
-        if (owner.equals(outerClass) || outerClass.type().isSubtypeOf(ownerType)) {
+  private static class MethodReference {
+
+    private final Symbol.MethodSymbol methodSymbol;
+    private final Symbol methodScopeOwner;
+    private boolean nonStaticReference = false;
+
+    MethodReference(Symbol.MethodSymbol symbol) {
+      methodSymbol = symbol;
+      methodScopeOwner = scopeOwner(methodSymbol);
+      if (methodScopeOwner != null && methodScopeOwner.isTypeSymbol()) {
+        nonStaticReference = !methodScopeOwner.isStatic() && !methodScopeOwner.owner().isPackageSymbol();
+      }
+    }
+
+    @CheckForNull
+    private static Symbol scopeOwner(Symbol symbol) {
+      Symbol owner = symbol.owner();
+      while (owner != null) {
+        if (owner.isPackageSymbol() || owner.isTypeSymbol() || owner.isMethodSymbol()) {
+          break;
+        }
+        owner = owner.owner();
+      }
+      return owner;
+    }
+
+    @CheckForNull
+    private static Symbol getPackage(Symbol symbol) {
+      Symbol owner = symbol.owner();
+      while (owner != null) {
+        if (owner.isPackageSymbol()) {
+          break;
+        }
+        owner = owner.owner();
+      }
+      return owner;
+    }
+
+    void setNonStaticReference() {
+      nonStaticReference = true;
+    }
+
+    boolean hasNonStaticReference() {
+      return nonStaticReference;
+    }
+
+    void checkSymbol(Symbol symbol) {
+      if (nonStaticReference || symbol == methodSymbol || symbol.isStatic()) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)